src/vintch_model/subunit_model.py

```diff
@@ -0,0 +1,178 @@
+            )  # which time to take
+        else:
+            self._pooling_weights = self.backend_class.randn((n_channels, n_dim, n_dim))
+        self.pooling_in_time = pooling_in_time
+        self._pooling_biases = self.backend_class.randn((n_channels, 1))
+        self.nonlinearities_chn = [
+            TentNonlinearity(n_basis_funcs, soft=True, backend=self.backend_class.name)
+            for _ in range(n_channels)
+        ]
+        self.nonlinearity_out = TentNonlinearity(
+            n_basis_funcs, soft=True, backend=self.backend_class.name
+        )
+
+        if kernels is not None:
+            self.kernels = kernels
+        if pooling_weights is not None:
+            self.pooling_weights = pooling_weights
+        if pooling_biases is not None:
+            self.pooling_biases = pooling_biases
+
+    def __call__(self, *args, **kwds):
+        return self.forward(*args, **kwds)
+
+    @property
+    def kernels(self):
+        return self._kernels
+
+    @kernels.setter
+    def kernels(self, new_kernels: Any):
+        """Setter for kernel filters."""
+        assert (
+            new_kernels.shape == self._kernels.shape
+        ), f"Expected kernel shape to be {self._kernels.shape}, got {new_kernels.shape} instead."
+        self.backend_class.check_input_type(new_kernels)
+        self._kernels = new_kernels
+
+    @property
+    def pooling_weights(self):
+        return self._pooling_weights
+
+    @pooling_weights.setter
+    def pooling_weights(self, weights: Any):
+        """Setter for pooling weights."""
+        assert (
+            weights.shape == self._pooling_weights.shape
+        ), f"Expected pooling weights shape to be {self._pooling_weights.shape}, got {weights.shape} instead."
+        self.backend_class.check_input_type(weights)
+        self._pooling_weights = weights
+
+    @property
+    def pooling_biases(self):
+        return self._pooling_biases
+
+    @pooling_biases.setter
+    def pooling_biases(self, biases: Any):
+        """Setter for pooling biases."""
+        assert (
+            biases.shape == self._pooling_biases.shape
+        ), f"Expected pooling biases shape to be {self._pooling_biases.shape}, got {biases.shape} instead."
+        self.backend_class.check_input_type(biases)
+        self._pooling_biases = biases
+
+    def forward(self, x):
+        """
+        Forward pass through the whole model.
+        """
+        self.backend_class.check_input_type(x)
+        # ensure the input is 5D and the second dimension is 1 (grayscale)
+        if x.ndim != 5 or x.shape[1] != 1:
+            raise ValueError(
+                f"Input must be a 5D tensor with shape [batch, {self.n_channels}, time, height, width], got {x.shape} instead."
+            )
+        generator_signal = self.generator_forward(x)  # [batch, time]
+        out = self.nonlinearity_out.transform(generator_signal)  # [batch, time]
+        return out
+
+    def generator_forward(self, x):
```

Comment:
generator is an odd name. I know they call it "the generator signal" in the paper, but [generator](https://wiki.python.org/moin/Generators) means something specific in python. so maybe call it "generator signal" instead?

and add a docstring here which explains this is gets sent through the final nonlinearity in order to be turned into the cell's firing rate.